### PR TITLE
Switch `RecentEra era` to use ledger eras

### PR DESCRIPTION
**Pull request by Anviking**
### Issue Number

ADP-2353

### CODE REVIEW

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs [hunk 1]
```diff
@@ -776,35 +765,77 @@ toCardanoApiUTxO = withConstraints (recentEra @era) $
...
+type PParams = Core.PParams
+
+data PParamsInAnyRecentEra where
+    PParamsInAnyRecentEra
+        :: IsRecentEra era
+        => RecentEra era
+        -> Core.PParams era
+        -> PParamsInAnyRecentEra
+
+toRecentEraGADT
```
Comment (Anviking) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs:786]: Now that it is specialized to `PParams` one should perhaps have a better name 🤔

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs [hunk 2]
```diff
@@ -601,7 +578,7 @@ data TxOutInRecentEra =
 unwrapTxOutInRecentEra
     :: RecentEra era
     -> TxOutInRecentEra
-    -> (TxOut (CardanoApi.ShelleyLedgerEra era))
+    -> (TxOut era)
```
Comment (Anviking) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs]:
```suggestion
    -> TxOut era
```

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs [hunk 3]
```diff
@@ -384,39 +388,44 @@ toRecentEra = \case
...
-shelleyBasedEraFromRecentEra :: RecentEra era -> CardanoApi.ShelleyBasedEra era
+shelleyBasedEraFromRecentEra
+    :: RecentEra era
+    -> CardanoApi.ShelleyBasedEra (CardanoApiEra era)
 shelleyBasedEraFromRecentEra = \case
     RecentEraConway -> CardanoApi.ShelleyBasedEraConway
     RecentEraBabbage -> CardanoApi.ShelleyBasedEraBabbage
 
--- | For convenience working with 'IsRecentEra'.
---
```
Comment (Anviking) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs:409]: Looks like a bad rebase

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs [hunk 4]
```diff
@@ -384,39 +388,44 @@ toRecentEra = \case
...
+-- | For convenience working with 'IsRecentEra'. Similar to 'Cardano.cardanoEra,
+-- but with a 'IsRecentEra era' constraint instead of 'Cardano.IsCardanoEra.
+cardanoEra
+    :: forall era. IsRecentEra era
+    => CardanoApi.CardanoEra (CardanoApiEra era)
 cardanoEra = cardanoEraFromRecentEra $ recentEra @era
 
--- | For convenience working with 'IsRecentEra'.
---
--- Similar to 'CardanoApi.shelleyBasedEra, but with a 'IsRecentEra era'
```
Comment (Anviking) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs:417]: Looks like a bad rebase

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 5]
```diff
@@ -2665,9 +2664,11 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
     withWorkerCtx api walletId liftE liftE $ \wrk -> do
         let db = wrk ^. dbLayer
             netLayer = wrk ^. networkLayer
+
+            trWorker :: Tracer IO W.WalletLog
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2668]: Unfortunately we need this again — and I don't know why.

- Removing the `CardanoApiEra` superclass constraints didn't help
- Moving `trWorker` to after `W.readNodeTipStateForTxWrite` helps
- My attempt at reproducing with a smaller example failed https://cardanofoundation.slack.com/archives/C05G0PE6SKS/p1700239260332509?thread_ts=1700237492.919709&cid=C05G0PE6SKS

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 6]
```diff
@@ -3171,13 +3172,15 @@ constructSharedTransaction
     delegationRequest <-
         liftHandler $ traverse parseDelegationRequest $ body ^. #delegations
 
-    withWorkerCtx api wid liftE liftE $ \wrk -> do
+    withWorkerCtx @_ @_ @Handler api wid liftE liftE $ \wrk -> do
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:3175]:
```suggestion
    withWorkerCtx api wid liftE liftE $ \wrk -> do
```

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 7]
```diff
@@ -14,14 +14,14 @@
...
 {-# LANGUAGE TypeApplications #-}
-{-# LANGUAGE UndecidableInstances #-}
 
 {- HLINT ignore "Use ||" -}
 
 -- TODO: https://cardanofoundation.atlassian.net/browse/ADP-2841
 {-# LANGUAGE CPP #-}
 #if __GLASGOW_HASKELL__ >= 902
 {-# OPTIONS_GHC -fno-warn-ambiguous-fields #-}
+{-# LANGUAGE UndecidableInstances #-}
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Is the change of location deliberate? 😄

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 8]
```diff
@@ -539,9 +526,9 @@ balanceTransaction
         , IsRecentEra era
         )
     => UTxOAssumptions
-    -> ProtocolParameters era
-    -- ^ 'CardanoApi.ProtocolParameters' can be retrieved via a Local State
-    -- Query to a local node.
+    -> PParams era
+    -- ^ 'Cardano.ProtocolParameters' can be retrieved via a Local State Query
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: What does the `Cardano` namespace refer to, exactly?

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 5 above)
Comment (jonathanknowles) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2668]: @Anviking I can empathise with your frustration!

In this particular case, at least we can say that:
- the type annotation serves as documentation
- it doesn't harm readability

With that in mind, I think we probably don't need to worry too much (for the moment).

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs [hunk 9]
```diff
@@ -351,30 +359,26 @@ type RecentEraLedgerConstraints era =
...
 -- 'RecentEra'.
---
--- TODO [ADP-2354] Make 'RecentEraLedgerConstraints' superclass of
--- 'IsRecentEra'. Currently this would cause weird type-inference errors in the
--- wallet code specifically with GHC 8.10.
--- https://cardanofoundation.atlassian.net/browse/ADP-2353
 withConstraints
-    :: RecentEra era -> (RecentEraConstraints era => a) -> a
+    :: RecentEra era
+    -> ( (IsRecentEra era, CardanoApi.ShelleyLedgerEra (CardanoApiEra era) ~ era ) => a) -> a
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs]: See bcb758feadf397d0168ba67d0294ff972772d3ba (from #4252).

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 10]
```diff
@@ -1610,7 +1598,7 @@ coinSelectionErrorToBalanceTxError era = withConstraints era $ \case
         ErrBalanceTxInsufficientCollateral
         ErrBalanceTxInsufficientCollateralError
             { largestCombinationAvailable
-                = largestCombinationAvailable
+                = withConstraints era $ largestCombinationAvailable
                 & fmap W.TokenBundle.fromCoin
                 & toExternalUTxOMap
                 & fromWalletUTxO era
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Suggestion: avoid mixing `$` and `&`: mixing these means that the code can no longer be read purely from left-to-right or from right-to-left.

I've made a fix for this here: e009054135ab30b555e08ba05c5f7f5a7588a351.

**Review by jonathanknowles — COMMENTED** [review]
Hi @Anviking 

Thanks very much for making this PR -- it's certainly a relief to see that we can finally simplify `ShelleyLedgerEra era` to just `era` in so many places!

I left a few comments, but I've also made a a few suggested changes in a PR here: #4252

Main points:
- I think it's possible to simplify the type signature of `withConstraints`.
   Fixed in bcb758feadf397d0168ba67d0294ff972772d3ba
- This PR deviates from our coding standard in a few places.
   Fixed in f7b5bb908855d2a1af7acb72b62e7d9da040d2fe

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs (hunk 8 above)
Comment (Anviking) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: Hm... This seems like a problematic rebase on my part, thought I fixed these, will have a another look...

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs (hunk 10 above)
Comment (Anviking) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: However, this `withConstraints` should now be avoidable... (5afae154cfb544abe1714db1be85d56b31ba8337) I'll see what other `withConstraints` calls I can remove...

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs (hunk 7 above)
Comment (Anviking) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: No — and it's also not needed anymore. Dropped.

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs (hunk 8 above)
Comment (Anviking) [reply lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs]: [3b673c2](https://github.com/cardano-foundation/cardano-wallet/pull/4242/commits/3b673c267565d657d25f57b9a7844cba15cf7f7c)

**Review by jonathanknowles — APPROVED** [review]

**Pull request merged**